File: servidor/servidor.go
package main

import(
	"fmt"
	//Sincronización de goroutines
	"sync"
	//Manejo de errores
	"errors"
	//Conexión en la red
	"net"
	//para usar el paquete de strings
	"strings"
	//Protocolo
	"encoding/json"
)

//Estructura que contenga todas las salas
type Servidor struct{
	nombre string
	salas map[string]*Sala
	usuarios map[string]*Cliente
	candado sync.RWMutex
}

//Regresa el nombre del servidor
func (s *Servidor) getNombre() string {
	return s.nombre
}

//Tiene que regresar una copia para evitar los bloqueos
func (s *Servidor) getSalas() map[string]*Sala {
	s.candado.RLock()
	defer s.candado.RUnlock()

	//Creamos la copia
	copia := make(map[string]*Sala)
	for nombre, sala := range s.salas{
		copia[nombre] = sala
	}
	
	//Regresamos la copia
	return copia
}

//Tiene que regresar una copia para evitar los bloqueos
func (s *Servidor) getUsuarios() map[string]*Cliente {
	s.candado.RLock()
	defer s.candado.RUnlock()
	//Creamos la copia
	copia := make(map[string]*Cliente)
	for nombre, usuario := range s.usuarios{
		copia[nombre] = usuario
	}
	
	//Regresamos la copia
	return copia
}


//Crea el servidor
func NuevoServidor(s string) *Servidor {
	return &Servidor{
		nombre: s,
		salas: make(map[string]*Sala),
		usuarios: make(map[string]*Cliente),
	}
}

//Metodo para crear y agregar salas al servidor
func (s *Servidor) NuevaSala(nombre string) (*Sala, error){
	s.candado.Lock()
	defer s.candado.Unlock()

	_, existe := s.salas[nombre]
	if existe {
		return nil, errors.New("El nombre de la sala ya existe")
	}

	nuevaSala := &Sala{
		nombre: nombre,
		clientes: make(map[string]*Cliente),
		invitados: make(map[string]*Cliente),
	}

	s.salas[nombre] = nuevaSala

	return nuevaSala, nil	
}

//Fucnión par eliminar salas, unicamente se llama con salas existentes
func (s *Servidor) EliminaSala(sala *Sala){
	s.candado.Lock()
	defer s.candado.Unlock()

	//Elimina la sala del servidor
	delete(s.salas, sala.getNombre())
	return	
}

//Metodo para agregar un usuario al servidor
func (s *Servidor) NuevoCliente(nombre string, conexion net.Conn, c *json.Encoder, d *json.Decoder) (*Cliente, error) {
	s.candado.Lock()
	defer s.candado.Unlock()

	_, existe := s.usuarios[nombre]
	if existe{
		return nil, errors.New("El nombre de usuario ya existe")
	}

	nuevoCliente := &Cliente{
		conexion: conexion,
		usuario: nombre,
		estado: ACTIVE,
		codificador: c,
		decodificador: d,
	}
	s.usuarios[nombre] = nuevoCliente
	return nuevoCliente, nil
}

//Metodo para eliminar un cliente
func (s *Servidor) EliminaCliente(cliente *Cliente){

	//Elimina al usuario del servidor
	delete(s.usuarios, cliente.getUsuario())

	//Envia mensje de usuario desconectado a todos los clientes
	mt := FabricaMensaje(DISCONNECTED).username(cliente.getUsuario())
	s.EnviaTodos(cliente, nil, mt)
	
	
	//Verifica si el usuario estaba en alguna sala
	for _, sala := range s.getSalas(){
		if sala.Existe(cliente) {
			//Si era el unico en la sala elimina la sala
			if len(sala.clientes) == 1 {
				s.EliminaSala(sala)
			}else{
				//Si no manda llamar el metodo correspondiente
				sala.EliminaCliente(cliente)
			}
		}
	}
}

func (s *Servidor) Temporal(conexion net.Conn){

	//Decodificador y codificador para que cada cliente pueda leer y escribir en su propio hilo
	decodificador := json.NewDecoder(conexion)
	codificador := json.NewEncoder(conexion)

	//Creamos la variable que va a guardar el json
	var identificacion Mensaje
	err := decodificador.Decode(&identificacion)
	if err != nil{
		//Manda mensaje de Json invalido
		codificador.Encode(Mensaje{Type: RESPONSE, Operation: INVALID_TIPO, Result: INVALID_RESPUESTA})
		//Desconecta al cliente
		conexion.Close()
		return	
	}

	//Nombre que se quiere poner el usuario
	nombre := strings.TrimSpace(identificacion.Username)

	if identificacion.Type != IDENTIFY || nombre == "" || len(nombre) > 8 {
		//Manda mensaje de usuario no identificado
		codificador.Encode(Mensaje{Type: RESPONSE, Operation: INVALID_TIPO, Result: NOT_IDENTIFIED})
		//Desconecta al cliente
		conexion.Close()
		return
	}

	//Manda llamar el metodo NuevoCliente para agregarlo a la lista de usuarios del servidor
	c, e := s.NuevoCliente(nombre, conexion, codificador, decodificador)
	if e != nil {
		//Manda mensaje de usuario repetido
		codificador.Encode(Mensaje{Type: RESPONSE, Operation: IDENTIFY, Result: USER_ALREADY_EXISTS, Extra: nombre })
		//Desconecta al cliente
		conexion.Close()
		return
	}

	//Tranforma el mensaje en un json para poderlo imprimir en el servidor
	json, _ := json.Marshal(identificacion)
	fmt.Println("Servidor <<", c.getUsuario(), "---", string(json))

	//Crea y envia el mensaje correspondiente
	mc := FabricaMensaje(RESPONSE).operation(IDENTIFY).result(SUCCESS).extra(nombre)
	mt := FabricaMensaje(NEW_USER).username(nombre)
	s.EnviaTodos(c, mc, mt)

	//Lamar al ciclo infinito del cliente
	s.EscuchaCliente(c)
	
}

//Ciclo para que se escuchen todos los mensajes de un cliente
func (s *Servidor) EscuchaCliente(c *Cliente){
	//Desconecta al cliente
	defer s.Desconecta(c)
	for{
		var mensaje Mensaje
		err := c.decodificador.Decode(&mensaje)
		if err != nil {
			//Manda mensaje de Json invalido
			c.codificador.Encode(Mensaje{Type: RESPONSE, Operation: INVALID_TIPO, Result: INVALID_RESPUESTA})
			return
		}
		
		//Convierte el mensaje en un json para poderlo imprimir en el servidor
		json, _ := json.Marshal(mensaje)
		fmt.Println("Servidor <<", c.getUsuario(), "---", string(json))

		//Acción del mensaje
		s.ProcesaMensaje(mensaje, c)	
	}
	
}

//Define que hace el mensaje que envio el cliente
func (s *Servidor) ProcesaMensaje(m Mensaje, c *Cliente){
	switch m.Type {
	case IDENTIFY:
		//Deberia desconectar o solo ignorar?
		//s.Desconecta(c)
		
	case STATUS:
		//Revisa si el estado es distinto al que tiene el cliente
		if m.Status == c.getEstado(){
			return
		}
		//Actualiza el estado
		c.setEstado(m.Status)
		//No es posible que suceda un error 

		//Fabrica el mensaje
		mt := FabricaMensaje(NEW_STATUS).username(c.getUsuario()).status(m.Status)
		//Envia el mensaje
		s.EnviaTodos(c, nil, mt)
		
	case USERS:
		//Crea el map de los usuarios
		users := make(map[string]string)
		for n, cliente := range s.getUsuarios() {
			estado, _ := cliente.getEstado().toStringEstados()
			users[n] = estado
		}

		//Fabrica el mensaje y lo envia
		mc := FabricaMensaje(USER_LIST).users(users)
		s.EnviaTodos(c, mc, nil)
		
	case TEXT:
		//Checa si el usuario existe en el servidor
		destino , existe := s.usuarios[m.Username]
		if !existe{
			//Si no existe envia el mensaje correspondiente
			mc := FabricaMensaje(RESPONSE).operation(TEXT).result(NO_SUCH_USER).extra(m.Username)
			s.EnviaTodos(c, mc, nil)
			return
		}

		//Fabrica y envia el mensaje al destinatario
		md := FabricaMensaje(TEXT_FROM).username(c.getUsuario()).text(m.Text)
		s.EnviaTodos(destino, md, nil)

	case PUBLIC_TEXT:
		//Fabrica y envia el mensaje a todos
		mt := FabricaMensaje(PUBLIC_TEXT_FROM).username(c.getUsuario()).text(m.Text)
		s.EnviaTodos(c, nil, mt)

	case NEW_ROOM:
		//Verifica que el nombre de la sala tenga una longitud menor o igual a 16
		if len(m.Roomname) > 16{
			return
		}
		//Crea la sala
		sala, err := s.NuevaSala(m.Roomname)
		//Significa que la sala ya existe y manda el mensaje correspondiente
		if err != nil {
			mc := FabricaMensaje(RESPONSE).operation(NEW_ROOM).result(ROOM_ALREADY_EXISTS).extra(m.Roomname)
			s.EnviaTodos(c, mc, nil)
			return
		}

		//Agrega al cliente
		sala.AgregaCliente(c)
		//Manda el mensaje de confirmación
		mc := FabricaMensaje(RESPONSE).operation(NEW_ROOM).result(SUCCESS).extra(m.Roomname)
		s.EnviaTodos(c, mc, nil)
		
	case INVITE:
		//Verifica
		sala := s.Verifica(c, m, INVITE)
		if sala == nil {
			return
		}
		//Si el cliente no pertenece a la sala ignora el json
		if !sala.Existe(c){
			return
		}

		//Agrega al cliente en la lista de invitados y envia el mensaje
		for _, cliente := range m.Usernames {
			ct, _ := s.usuarios[cliente]
			sala.Invita(ct)
			ct.EnviaMensaje(FabricaMensaje(INVITATION).username(ct.getUsuario()).roomname(m.Roomname))
		}

	case JOIN_ROOM:
		//Verifica que la sala exista
		sala := s.Verifica(c, m, JOIN_ROOM)
		if sala == nil{
			return
		}
		
		//Agrega el usuario a la sala
		sala.AgregaCliente(c)

		//Fabrica y manda el mensaje correspondiente 
		mc := FabricaMensaje(RESPONSE).operation(JOIN_ROOM).result(SUCCESS).extra(m.Roomname)
		mt := FabricaMensaje(JOIN_ROOM).roomname(m.Roomname).username(c.getUsuario())
		sala.EnviaSala(c, mc, mt)

	case ROOM_USERS:
		//Verifica
		sala := s.Verifica(c, m, ROOM_USERS)
		if sala == nil{
			return
		}
		
		//Crea el map de los usuarios
		users := make(map[string]string)
		sala.candado.RLock()
		for n, cliente := range sala.getClientes() {
			estados, _ := cliente.getEstado().toStringEstados()
			users[n] = estados
		}
		sala.candado.RUnlock()

		//Fabrica y envia el mensaje
		mc := FabricaMensaje(ROOM_USERS_LIST).users(users)
		s.EnviaTodos(c, mc, nil)

	case ROOM_TEXT:
		//Verifica
		sala := s.Verifica(c, m, ROOM_TEXT)
		if sala == nil {
			return
		}

		//Fabrica y envia el mensaje a todos en la sala
		mt := FabricaMensaje(ROOM_TEXT_FROM).roomname(m.Roomname).username(c.getUsuario()).text(m.Text)
		sala.EnviaSala(c, nil, mt)

	case LEAVE_ROOM:
		//Verifica
		sala := s.Verifica(c, m, LEAVE_ROOM)
		if sala == nil{
			return	
		}

		//Si es el unico usuario en la sala elimina la sala
		if len(sala.getClientes()) == 1 {
			s.EliminaSala(sala)
		}else{
			//Si no, elimina al usuario de la sala
			sala.EliminaCliente(c)
		}
		
		//Por practicidad el mensaje lo envia el metodo EliminaCliente

	case DISCONNECT:
		//Manda a llamar el metodo desconecta
		s.Desconecta(c)
		
	default:
		//Nunca deberia pasar este default
		return
	}
	return
}

func (s *Servidor) Desconecta(c *Cliente){
	
	//Elimina al cliente de las salas
	for _, sala := range s.getSalas() {
		if len(sala.getClientes()) == 1 {
			s.EliminaSala(sala)
		} 
		sala.EliminaCliente(c)
	}
	
	//Elimina al cliente del servidor
	s.EliminaCliente(c)
	
	//Desconecta al usuario
	c.conexion.Close()
}

//Verifica que el usuario pueda hacer lo que quiere hacer en las salas
func (s *Servidor) Verifica(c *Cliente, m Mensaje, tipo Tipo) *Sala {
	//Verifica que la sala exista
	sala, existe := s.salas[m.Roomname]
	if !existe{
		mc := FabricaMensaje(RESPONSE).operation(tipo).result(NO_SUCH_ROOM).extra(m.Roomname)
		s.EnviaTodos(c, mc, nil)
		return nil
	}
	switch tipo {
	case LEAVE_ROOM, ROOM_TEXT, ROOM_USERS:
		//Verifica que el usuario este en el cuarto
		if !sala.Existe(c) {
			mc := FabricaMensaje(RESPONSE).operation(tipo).result(NOT_JOINED).extra(m.Roomname)
			s.EnviaTodos(c, mc, nil)
			return nil
		}
	case JOIN_ROOM:
		//Verifica que el usuario este invitado en el cuarto
		if !sala.Invitado(c) {
			mc := FabricaMensaje(RESPONSE).operation(tipo).result(NOT_INVITED).extra(m.Roomname)
			s.EnviaTodos(c, mc, nil)
			return nil
		}
	case INVITE:
		for _, invitado := range m.Usernames {
			_, e := s.usuarios[invitado]
			if ! e {
				mc := FabricaMensaje(RESPONSE).operation(INVITE).result(NO_SUCH_USER).extra(invitado)
				s.EnviaTodos(c, mc, nil)
				return nil
			}	
		}
		
	}
	return sala
}

func (s *Servidor) EnviaTodos(c *Cliente, mc *Mensaje, mt *Mensaje){

	//Si no hay mensaje al propio usuario
	if mc == nil{
		for usuario, cliente := range s.getUsuarios() {
			if usuario != c.getUsuario() {
				cliente.EnviaMensaje(mt)
			}
		}
	}else if mt == nil {
		//Si no hay mensaje para el resto de usuarios
		c.EnviaMensaje(mc)
	}else {
		//Hay mensaje para ambos
		for usuario, cliente := range s.getUsuarios() {
			if usuario == c.getUsuario() {
				cliente.EnviaMensaje(mc)
			}else{
				cliente.EnviaMensaje(mt)
			}
		}
	}
}

File: cliente/servidor.cs
using System;
using System.Collections.Generic;

public class Servidor {

    Cliente c;
    Dictionary<string, Estados> clientes;
    Dictionary<string, List<string>> salas;
    List<string> invitaciones;
    private object candado = new Object();

    //Metodo para crear un nuevo servidor
    public Servidor(Cliente c, Dictionary<string, Estados> clientes ) {
	this.c = c;
	this.clientes = clientes ?? new Dictionary<string, Estados>();
	this.salas = new Dictionary<string, List<string>>();
	this.invitaciones = new List<string>();
    }

    //Metodo para unirse a una sala
    public void UnirseSala(string nombre){
	Mensaje cs = Mensaje.FabricaMensaje(Tipo.ROOM_USERS);
	c.Envia(cs);
	return;
    }

    //Metodo para interpretar los mensajes del cliente y convertirlos en un json
    
    public void Traductor(string s){
	
	string[] separada = s.Split('-');

	//Para que no afecten los espacios antes o despues
	for(int i = 0; i<separada.Length; i++)
	    separada[i] = separada[i].Trim();
	
	switch (separada[0]){
	    case "Actualiza estado":
	    case "actualiza estado":
	    case "ACTUALIZA ESTADO":
	    if(separada[1] == ("ACTIVE") || separada[1] == ("active") || separada[1] ==("Active")){
		Mensaje me = Mensaje.FabricaMensaje(Tipo.STATUS).status(Estados.ACTIVE);
		c.Envia(me);
	    }else if(separada[1] == ("AWAY") || separada[1] == ("away") || separada[1] == ("Away")){
		Mensaje me = Mensaje.FabricaMensaje(Tipo.STATUS).status(Estados.AWAY);
		c.Envia(me);
	    }else if(separada[1] == ("BUSY") || separada[1] == ("busy") || separada[1] == ("Busy")){
		Mensaje me = Mensaje.FabricaMensaje(Tipo.STATUS).status(Estados.BUSY);
		c.Envia(me);
	    }else
		Console.WriteLine("El estado no es válido");
	    break;
	    case "usuarios":
	    case "Usuarios":
	    case "USUARIOS":
		Console.WriteLine("Clientes " + " - " + " Estados");
		if(separada[1] == "General")
		    foreach(var cliente in clientes)
			
			Console.WriteLine(cliente.Key + " - " + cliente.Value);
		else{
		    foreach(var sala in salas)
			if(sala.Key == separada[1])
			    foreach(var cliente in sala.Value){
				Estados e = clientes[cliente];
				Console.WriteLine(cliente + "-" + e );
			    }
		    
		    Console.WriteLine("Es necesario pertenecer a la sala para acceder a la lista de usuarios");
		}		
		break;
	    case "Crea":
	    case "CREA":
	    case "crea":
		Mensaje mc = Mensaje.FabricaMensaje(Tipo.NEW_ROOM).roomname(separada[1]);
		c.Envia(mc);
		break;
	    case "Invita":
	    case "INVITA":
	    case "invita":
		string[] invitados = separada[1].Split(',');
		foreach(string invitado in invitados)
		    if (!clientes.ContainsKey(invitado)){
			Console.WriteLine("El usuario "+ invitado+ " no existe");
			break;
		    }
		Mensaje mi = Mensaje.FabricaMensaje(Tipo.INVITE).roomname(separada[1]).usernames(invitados);
		c.Envia(mi);
		break;
	    case "UNIRSE":
	    case "Unirse":
	    case "unirse":
		foreach(string invitado in invitaciones)
		    if(invitado == separada[1]){
			Mensaje mu = Mensaje.FabricaMensaje(Tipo.JOIN_ROOM).roomname(separada[1]);
			c.Envia(mu);
			break;
		    }
		Console.Write("Es necesario que estes invitado a la sala para unirtea la sala " + separada[1]);
		break;
	    case "ABANDONA":
	    case "Abandona":
	    case "abandona":
		foreach(string sala in salas.Keys)
		    if(sala == separada[1]){
			Mensaje ma = Mensaje.FabricaMensaje(Tipo.LEAVE_ROOM).roomname(separada[1]);
			c.Envia(ma);
			break;
		    }
		Console.WriteLine("No eres miembro de la sala "+separada[1]);
		break;
	    case "DESCONECTA":
	    case "Desconecta":
	    case "desconecta":
	    Desconecta();
	    break;
	    default:
		if (separada.Length == 2)
		    if(separada[0] == "TODOS" || separada[0] == "Todos" || separada[0] == "todos"){
			Mensaje mt = Mensaje.FabricaMensaje(Tipo.PUBLIC_TEXT).text(separada[1]);
			c.Envia(mt);
		    }else {
			foreach (string usuario in clientes.Keys)
			    if (usuario == separada[0]){
				Mensaje mu = Mensaje.FabricaMensaje(Tipo.TEXT).username(usuario).text(separada[1]);
				c.Envia(mu);
				break;
			    }
			
			foreach (string sala in salas.Keys)
			    if(sala == separada[0]){
				Mensaje ms = Mensaje.FabricaMensaje(Tipo.ROOM_TEXT).roomname(separada[1]);
				c.Envia(ms);
				break;
			    }
			//No se encontro el usuario o sala
			Console.WriteLine("El usuario o la sala "+ separada[1]+ " no existe");
		    }
		else{
		    Console.WriteLine("El formato no coincide con el uso del chat");
		    UsoServidor();
		}
		break;	    
	}
	return;
    }

    //Escucha los mensajes del servidor
    public void EscuchaServidor(){
	//Mientras siga conectado con el servidor
	while (true) {
	    //Decodifica el mensaje
	    Mensaje m = c.Recibe();

	    //Si el mensaje no es valido
	    if(m == null)
		return;

	    switch (m.Type){
		case Tipo.RESPONSE:
		    ManejaRespuestas(m);
		    break;
		case Tipo.NEW_USER:
		    lock(candado){
			clientes.Add(m.Username, Estados.ACTIVE);
		    }
		    break;
		case Tipo.NEW_STATUS:
		    lock(candado){
			clientes[m.Username] = m.Status;
		    }
		    break;
		case Tipo.TEXT_FROM:
		    Console.WriteLine(m.Username + "- " + m.Text);
		    break;
		case Tipo.PUBLIC_TEXT_FROM:
		    Console.WriteLine("General - "+ m.Username +" - " + m.Text);
		    break;
		case Tipo.INVITATION:
		    Console.WriteLine("El usuario " + m.Username + " te ha invitado a la sala "+ m.Roomname);
		    invitaciones.Add(m.Roomname);
		    break;
		case Tipo.JOINED_ROOM:
		    salas[m.Roomname].Add(m.Username);
		    break;
		case Tipo.ROOM_USERS_LIST:
		    salas.Add(m.Roomname, new List<string>(m.Users.Keys));
		    salas[m.Roomname].Add(c.Username);
		    break;
		case Tipo.ROOM_TEXT_FROM:
		    Console.Write(m.Roomname + "-" + m.Username + ":" + m.Text);
		    break;
		case Tipo.LEFT_ROOM:
		    salas[m.Roomname].Remove(m.Username);
		    break;
		case Tipo.DISCONNECTED:
		    lock(candado){
			clientes.Remove(m.Username);
		    }
		    break;
		default:
		    Console.WriteLine("Error al recibir la respuesta");
		    Desconecta();
		    break;
	    }
	}
    }

    public void ManejaRespuestas(Mensaje m){
	switch (m.Result){
	    case Respuesta.SUCCESS:
		ManejaOperacion(m);
		break;
	    case Respuesta.NO_SUCH_USER:
		Console.WriteLine("El usuario "+ m.Extra + " no existe");
		break;
	    case Respuesta.ROOM_ALREADY_EXISTS:
		Console.WriteLine("El nombre de sala " + m.Extra + " ya existe en el servidor");
		break;
	    case Respuesta.NO_SUCH_ROOM:
		Console.WriteLine("La sala "+ m.Extra + " no existe");
		break;
	    case Respuesta.NOT_INVITED:
		Console.WriteLine("No fue posible unirse a la sala " + m.Extra + " debido a que no has sido invitado");
		break;
	    case Respuesta.NOT_JOINED:
		Console.WriteLine("No formas parte de la sala " + m.Extra);
		break;
	    default:
		Console.WriteLine("Algo salio mal");
		Desconecta();
		break;
	}
	return;
    }

    public void ManejaOperacion(Mensaje m){
	switch(m.Operation){
	    case Tipo.NEW_ROOM:    
		salas.Add(m.Extra, new List<string> {c.Username});
		Console.WriteLine("La sala " + m.Extra + " ha sido creada correctamente");
		break;
	    case Tipo.JOIN_ROOM:
		UnirseSala(m.Extra);
		invitaciones.Remove(m.Extra);
		break;
	    default:
		Console.WriteLine("Ocurrio un error inesperado");
		Desconecta();
		break;
		}
	return;
    }

    //Metodo que explica como usar el chat
    public void UsoServidor(){
	Console.WriteLine("Uso del chat:\n" +
			  "Para actalizar un estado escriba AWAY, BUSY o ACTIVE:\n " +
			  "                        Actaliza estado - estado\n" +
			  "Para obtener la lista de usuarios del servidor escriba:\n" +
			  "                        Usuarios - General\n" +
			  "Para mandar un mensaje a todos escriba:\n" +
			  "                        Todos - mensaje\n" +
			  "Para mandar un mensaje a un usuario escriba\n" +
			  "                        nombre_del_usuario - mensaje\n" +
			  "Para mandar mensaje en una sala escriba\n" +
			  "                        nombre_de_la_sala - mensaje\n"+
			  "Para crear una sala escriba:\n" +
			  "                        Crea - nombre_de_la_sala\n" +
			  "Para invitar a usuarios a la sala escriba:\n" +
			  "                        Invita - usuario1, usuario2, usuario3\n" +
			  "Para unirse a una sala a la que fue invitado escriba\n" +
			  "                        Unirse - nombre_de_la_sala\n" +
			  "Para obtener la lista de usuarios de una sala en especifico escrba\n" +
			  "                        Usuarios - nombre_de_la_sala\n" +
			  "Para abandonar una sala escriba:\n"+
			  "                        Abandona - nombre_de_la_sala\n" +
			  "Para desconectarse del servidor escriba:\n"+
			  "                        Desconecta");
	return;
    }

    //Metodo para enviar el mensaje de desconección  a los usuarios
    public void Desconecta(){
	Mensaje m = Mensaje.FabricaMensaje(Tipo.DISCONNECT);
	c.Envia(m);
	Environment.Exit(0);
	return;
    }
}
